From fa62d548c4edefd48b9287d7c82fdf9ee31dac10 Mon Sep 17 00:00:00 2001
From: Dmitry Ivanov <idmitry@qti.qualcomm.com>
Date: Wed, 9 Oct 2024 15:21:04 +0200
Subject: Unconditionally use glClearBufferXXX() instead of glClear()

This fixes the dEQP-GLES3.functional.fragment_ops.scissor#clear_uint_buffer
test.

The test enables a scissors and crears a framebuffer.  The result is compared to
a reference image.  The reference image is created by clearing all the framebuffer
with scissors disabled and then fixing the area that should not be cleared.

Test image is ok, the problem is in the reference image.  Those two
use different paths in mesa3d/virglrenderer.  To clear image with scissors,
mesa3d instead draws a quad with the clear color.  When mesa3d is requested
to clear all the framebuffer, it passes the corresponding command
to virglrenderer and the control arrives at this code.

In some cases the code calls a correct version of glClearBufferuXX(),
depending of the buffer format.  Often it just sets the color with glClearColor()
and then calls glClear(GL_STENCIL_BUFFER_BIT).  But glClearColor expects
the color arguments to be floats, and the buffer in the case of the test is
of type unsigned int.

This patch removes the call to glClear() to clear color buffers,
any clear color buffer call is now going through glClearBufferuXX().
The patch is not complete:

1. vrend_clear_prepare() still sets the colors with glClearColor() and
   vrend_clear_finish() still restores them.  Both of these functions
   are used also in vrend_clear_surface() below.

2. Probably we should remove all the calls to glClearColor()/glClear()
   through all the renderer in favor of glClearBufferuXX().

Signed-off-by: Dmitry Ivanov <idmitry@qti.qualcomm.com>
---
diff --git a/src/vrend_renderer.c b/src/vrend_renderer.c
index d0b7784f..041a03e1 100644
--- a/src/vrend_renderer.c
+++ b/src/vrend_renderer.c
@@ -4852,29 +4877,21 @@ void vrend_clear(struct vrend_context *ctx, unsigned buffers,
                        buffers, colorf, depth, stencil);
 
    if (buffers & PIPE_CLEAR_COLOR) {
-      uint32_t mask = 0;
-      int i;
-      for (i = 0; i < sub_ctx->nr_cbufs; i++) {
-         if (sub_ctx->surf[i])
-            mask |= (1 << i);
+      uint32_t mask = (buffers >> 2) & ((1 << PIPE_MAX_COLOR_BUFS) - 1);
+
+      while (mask) {
+         int i = u_bit_scan(&mask);
+
+         if (!sub_ctx->surf[i])
+            continue;
+
+         if (util_format_is_pure_uint(sub_ctx->surf[i]->format))
+            glClearBufferuiv(GL_COLOR, i, (GLuint *)colorf);
+         else if (util_format_is_pure_sint(sub_ctx->surf[i]->format))
+            glClearBufferiv(GL_COLOR, i, (GLint *)colorf);
+         else
+            glClearBufferfv(GL_COLOR, i, (GLfloat *)colorf);
       }
-      if (mask != (buffers >> 2)) {
-         mask = buffers >> 2;
-         while (mask) {
-            i = u_bit_scan(&mask);
-            if (i < PIPE_MAX_COLOR_BUFS && sub_ctx->surf[i] &&
-                util_format_is_pure_uint(sub_ctx->surf[i] &&
-                                         sub_ctx->surf[i]->format))
-                glClearBufferuiv(GL_COLOR, i, (GLuint *)colorf);
-            else if (i < PIPE_MAX_COLOR_BUFS && sub_ctx->surf[i] &&
-                     util_format_is_pure_sint(sub_ctx->surf[i] &&
-                                              sub_ctx->surf[i]->format))
-                glClearBufferiv(GL_COLOR, i, (GLint *)colorf);
-            else
-                glClearBufferfv(GL_COLOR, i, (GLfloat *)colorf);
-         }
-      } else
-         bits |= GL_COLOR_BUFFER_BIT;
    }
    if (buffers & PIPE_CLEAR_DEPTH)
       bits |= GL_DEPTH_BUFFER_BIT;
